src/llmq/signing_shares.h

```diff
@@ -376,20 +376,25 @@ class CSigSharesManager : public CRecoveredSigsListener
     static constexpr int64_t MAX_SEND_FOR_RECOVERY_TIMEOUT{10000};
     static constexpr size_t MAX_MSGS_SIG_SHARES{32};
 
-    RecursiveMutex cs;
-
     std::thread workThread;
     CThreadInterrupt workInterrupt;
 
-    SigShareMap<CSigShare> sigShares GUARDED_BY(cs);
-    std::unordered_map<uint256, CSignedSession, StaticSaltedHasher> signedSessions GUARDED_BY(cs);
+    Mutex cs_sigShares;
```

Comment:
it became a bit over-complicated IMO, 7 mutexes in just one class. Does it indeed improve performance by any benchmark? Maybe you can localize some specific data that affect performance and put only that one by its own mutex?


It looks like DEADLOCKS ARE WELCOME every time you change something a bit

Reply:
Bitcoin has tried to move away from using recursive mutexes as much as possible; and generally we've tried to follow. I think if you tried to do this with 1 non-recursive mutex, you'd have a very bad time. Instead we introduce mutex for the actual data structures they need to guard, and use annotations to give us decent assurance there are no issues

Reply:
Yes, in theory it sounds good, but how to proof validness of each piece of data, if there no locks? Is there any dependency between members? Is it true that each piece of member can be changed without side effect to other calculations in neighbors function?

Maybe it will work, but that's a nightmare to proof that there is jo hidden dependencies between members and review that we can't get invalid state now. 

Yes, bitcoin is moving away from RecursiveMutex, but it's not like "give own mutex to each variable in the header", at least most times

Reply:
maybe collect logs `Enter: lock contention.... started / completed` at least from some node to narrow down scope?

Reply:
I think I agree... Please see 6417 as a much more minimal alternative